src/tracked_struct.rs

```diff
@@ -396,12 +387,18 @@ where
         let current_revision = zalsa.current_revision();
         match zalsa_local.tracked_struct_id(&identity) {
             Some(id) => {
+                // The struct already exists in the intern map.
                 let index = self.database_key_index(id);
                 tracing::trace!("Reuse tracked struct {id:?}", id = index);
-                // The struct already exists in the intern map.
                 zalsa_local.add_output(index);
-                self.update(zalsa, current_revision, id, &current_deps, fields);
-                FromId::from_id(id)
+
+                let updated_id = self.update(zalsa, current_revision, id, &current_deps, fields);
+                if id != updated_id {
+                    // Overwrite the previous ID if we are reusing the slot with new fields.
+                    zalsa_local.store_tracked_struct_id(identity, updated_id);
```

Comment:
This feels a little wrong, but I think it is correct. If the struct is recreated with the same fields, we want to return the latest generation for the tracked structs to align with the memos. If it is recreated with new fields, we have to create a new generation and invalidate the previous version.